feat(controller): multiple TrafficRoutingReconciler

fblgit opened the pull request:
# Current Scenario
You can add multiple `trafficRouting` on rollouts spec. Unfortunately, only the first is used (alphabetically sorted). example:
```
    canary:
      canaryService: rollouts-demo-canary
      stableService: rollouts-demo-stable
      trafficRouting:
        smi: {}
        nginx:
          stableIngress: rollouts-demo-stable
```
In that case, only `nginx` will be used and `smi` will be ignored.

# Feature multiple TrafficRoutingReconciler
With the same spec:
```
    canary:
      canaryService: rollouts-demo-canary
      stableService: rollouts-demo-stable
      trafficRouting:
        smi: {}
        nginx:
          stableIngress: rollouts-demo-stable
```
In this case, both `trafficRouting` definitions are used and they are kept in sync as well. 
```
INFO[2021-09-04T14:01:19Z] Found 2 TrafficRouting Reconciliers           namespace=default rollout=rollouts-demo
INFO[2021-09-04T14:01:19Z] Reconciling TrafficRouting with type 'Nginx'  namespace=default rollout=rollouts-demo
INFO[2021-09-04T14:01:19Z] syncing service                               namespace=default rollout=rollouts-demo service=rollouts-demo-canary
INFO[2021-09-04T14:01:19Z] updating canary Ingress                       desiredWeight=5 ingress=rollouts-demo-rollouts-demo-stable-canary namespace=default rollout=rollouts-demo
INFO[2021-09-04T14:01:19Z] Event(v1.ObjectReference{Kind:"Rollout", Namespace:"default", Name:"rollouts-demo", UID:"02f3c5eb-18c1-41bb-ae05-bc80150c06b3", APIVersion:"argoproj.io/v1alpha1", ResourceVersion:"58651", FieldPath:""}): type: 'Normal' reason: 'PatchingCanaryIngress' Updating Ingress `rollouts-demo-rollouts-demo-stable-canary` to desiredWeight '5'
INFO[2021-09-04T14:01:19Z] Updating Ingress `rollouts-demo-rollouts-demo-stable-canary` to desiredWeight '5'  event_reason=PatchingCanaryIngress namespace=default rollout=rollouts-demo
INFO[2021-09-04T14:01:19Z] Desired weight (stepIdx: 0) 5 verified        namespace=default rollout=rollouts-demo
INFO[2021-09-04T14:01:19Z] Reconciling TrafficRouting with type 'SMI'    namespace=default rollout=rollouts-demo
INFO[2021-09-04T14:01:19Z] TrafficSplit `rollouts-demo` modified         event_reason=TrafficSplitModified namespace=default rollout=rollouts-demo
INFO[2021-09-04T14:01:19Z] Desired weight (stepIdx: 0) 5 verified        namespace=default rollout=rollouts-demo
INFO[2021-09-04T14:01:19Z] Event(v1.ObjectReference{Kind:"Rollout", Namespace:"default", Name:"rollouts-demo", UID:"02f3c5eb-18c1-41bb-ae05-bc80150c06b3", APIVersion:"argoproj.io/v1alpha1", ResourceVersion:"58651", FieldPath:""}): type: 'Normal' reason: 'TrafficSplitModified' TrafficSplit `rollouts-demo` modified
INFO[2021-09-04T14:01:19Z] Rollout step 1/7 completed (setWeight: 5)     event_reason=RolloutStepCompleted namespace=default rollout=rollouts-demo
```
This is performed with the minimum impact by just mutating the structure into a list that is processed one by one sequentially.

# Scenarios
A) existing fleet that can't transition from nginx to VirtualGateways or similars
B) adding sidecars to large throughput nginx controllers is not viable or can have a negative impact
C) adoption of full N-S and W-E canary with zero-downtime on large fleets
D) nginx-service-mesh (nginxinc) ++ ingress-nginx (kubernetes) doesn't work so well together
E) nginx service mesh N/S W/E without using nginxinc ingress.
F) no host header annotations to circumvent routing issues when adding nginx-ingress to a mesh

Despite facing both A+B+C+D+E+F Scenarios, there are a few thoughts behind this:
- Why have to choose between SMI or Nginx?
- Why change 3000 ingresses into VirtualGateways or similar?
- Why adding sidecars to the ingress if the annotations should work fine ?
- Why can accomplish either N/S or W/E traffic shift with minimal cost/impact ?
- Why multiple providers wouldn't be able to work together?

This will be the first to allow the use of multiple traffic-shift mechanisms among the other existing solutions that enables features such as canary around :)

Checklist:

* [ ] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-rollouts/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this is a chore.
* [X] The title of the PR is (a) [conventional](https://www.conventionalcommits.org/en/v1.0.0/), (b) states what changed, and (c) suffixes the related issues number. E.g. `"fix(controller): Updates such and such. Fixes #1234"`.  
* [X] I've signed my commits with [DCO](https://github.com/argoproj/argoproj)
* [X] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [X] My builds are green. Try syncing with master if they are not. 
* [ ] My organization is added to [USERS.md](https://github.com/argoproj/argo-rollouts/blob/master/USERS.md).

xavipanda commented:
We face similar scenarios, we built a custom controller for this to sync these two..

Hi @jessesuen
This was part of the milestone for argo-rollouts till not long ago, and now this contribution is at the door.
Do you have some time to review this please ? we would love to get rid of that controller and see rollouts taking care of both SMI and NGINX.

Appreciate your efforts! Thank you

harikrongali commented:
@xavipanda  
we will take a look at it and let you know

harikrongali commented:
@fblgit can you fix lint errors?

harikrongali commented:
@khhirani can you review the PR?

khhirani reviewed (changes requested):
@fblgit Can you please create new unit tests to verify that your change works? Right now your PR isn't meeting our code coverage requirement.

Please add some new E2E tests as well.

xavipanda commented:
sure, let me work on both lint and test-cases. thanks

fblgit commented:
sure, let me work on it :) thanks

fblgit commented:
Hi @khhirani 

e2e-tests:
* Added a new test scenario, SMI + Ingress: `make test-e2e E2E_TEST_OPTIONS="-testify.m ^TestSMIIngressCanaryStep$"`

Note: Added a few func's to get Ingress data. Let me know if u want a decoupled e2e just for the nginx, should be quickly doable.

lint:
Fixed sync.Mutex issue, now looking ok.
```
# make lint
go mod tidy
go mod vendor
golangci-lint run --fix
# 
```

unit-tests:
* added a unit-test for 3 reconcilers (alb + nginx + smi)
* added a unit-test for 2 reconcilers (nginx + smi)
* existing tests (networkReconcilers and trafficReconcilers) has been adapted to the new []type

coverage:
* should look better now, please run the workflow when u have some time 👍 

Hope next time I can nail the PR it at the first attempt, sorry for the trouble. Thank you

harikrongali commented on rollout/trafficrouting.go:
```diff
@@ -17,152 +17,177 @@ import (
-	if err != nil {
-		return err
-	}
+		var canaryHash, stableHash string
```
can you fix indentations? as it is unclear which of them changed.. this block isnt changed

fblgit replied:
this is some visual effect of github diff.
open the file in ur IDE, you won't see any indentation issue.
Line 97 adds a loop, where we have to iterate over the list of trafficReconcilers.
indent is right

harikrongali commented on rollout/trafficrouting.go:
```diff
@@ -17,152 +17,177 @@ import (
-				if step.SetWeight != nil {
-					desiredWeight = *step.SetWeight
-					break
+		currentStep, index := replicasetutil.GetCurrentCanaryStep(c.rollout)
```
even this block isnt changed..fix indentation

fblgit replied:
this is some visual effect of github diff.
open the file in ur IDE, you won't see any indentation issue.
Line 97 adds a loop, where we have to iterate over the list of trafficReconcilers.
indent is right

harikrongali commented on rollout/trafficrouting.go:
```diff
@@ -17,152 +17,177 @@ import (
-				for _, tmpl := range exStep.Templates {
-					if tmpl.Name == name {
-						return tmpl.Weight
+			// Checks for experiment step
```
same here

fblgit replied:
this is some visual effect of github diff.
open the file in ur IDE, you won't see any indentation issue.
Line 97 adds a loop, where we have to iterate over the list of trafficReconcilers.
indent is right

harikrongali commented on rollout/trafficrouting.go:
```diff
@@ -17,152 +17,177 @@ import (
-			c.enqueueRolloutAfter(c.rollout, 10*time.Second)
-		} else {
-			c.log.Infof("Desired weight (stepIdx: %d) %d verified", *index, desiredWeight)
+
```
same here.

fblgit replied:
this is some visual effect of github diff.
open the file in ur IDE, you won't see any indentation issue.
Line 97 adds a loop, where we have to iterate over the list of trafficReconcilers.
indent is right

harikrongali commented on test/e2e/smi/rollout-smi-experiment.yaml:
```diff
@@ -84,4 +84,4 @@ spec:
           requests:
             memory: 16Mi
-            cpu: 5m
+            cpu: 5m
```
no change. please revert this file

fblgit replied:
done

harikrongali commented:
@fblgit Thanks for the contribution. Can you fix the indentation issue?
Also can you add documentation as will with different scenarios and use cases. (provide some sample examples so that users can try).

fblgit commented:
@harikrongali
Happy to contribute to this great project!

Regarding your comments:

indent:
* it's a Github diff visual issue, please open the file in your IDE or `View File` in GH and you won't see any indent issue. check L97, there is a `for` that is required due to the change of []type

documentation:
* I did my best on this, but keep in mind I'm not an English native speaker so It would be very encouraging that someone looks at it and help on this element.
* clarified this will be available from v1.1 (as shown in the milestone for this PR)

example:
* Added a whole scenario, SMI+Ingress within the getting-started guide (SMI + Nginx)

Again thanks for your time, let me know if anything else is needed.

leoluz commented on docs/getting-started/mixed/index.md:
```diff
@@ -0,0 +1,232 @@
+
+![Rollout Paused](../nginx/paused-rollout-nginx.png)
+
+At this point, both the canary and stable version of the Rollout are running, with 5% of the
```
> At this point, both the canary and stable

Can you please confirm is this sentence is correct? It seems to me that what happens is that at this point both smi and nginx are receiving 5% of traffic in their canary pods?

xavipanda replied:
noted.

fblgit replied:
Interesting, I got it from `docs/getting-started/nginx/index.md` but your suggestion make sense.

fblgit replied:
ammended.

leoluz commented on docs/features/traffic-management/mixed.md:
```diff
@@ -0,0 +1,49 @@
+
+## When to use multiple providers
+
+When you do not want:
```
Is it possible to elaborate a bit more why it isn't recommended to use multiple providers in the cases bellow?

xavipanda replied:
noted

fblgit replied:
What about now, looks better? Is that to be honest, I don't think about a suggestion on where to use it or not.. but mostly about examples on when you can use it if that's what you want.

leoluz replied:
Nice. I understand much better what you mean by this section now. Tks!

leoluz commented on rollout/trafficrouting.go:
```diff
@@ -17,152 +17,177 @@ import (
 			ControllerKind: controllerKind,
 			IngressLister:  c.ingressesLister,
 		})
+		trafficReconcilers = append(trafficReconcilers, alb_reconcilier)
```
If `err != nil` you will be appending a `nil` `alb_reconciler` to the `trafficReconcilers` slice which may cause some negative side effect. It is better to check the error first and just append if everything is fine.

xavipanda replied:
I understand your point.
```
		return alb.NewReconciler(alb.ReconcilerConfig{
```
This is the original one, it was returning it in that way before.
Altering this could affect existing tests and behaviour.

I do agree with what you are saying here, there has to be a decision of how we want to action now onwards this multiple trafficRouting scenarios.
Originally, 1 provider could fail and `return nil, err`. So now I believe there are 3 choices:
1. return [valid, nil], err
2. return [nil], err
3. return [valid], err

I feel like number 2 could be a good approach, either you have a valid Rollout or not.. nothing in between.

fblgit replied:
in any case, the errors of the reconcilers are controlled in the `reconcileTrafficRouting()`. No matter what u append here if you have `err != nil` reconcileTrafficRouting will not process any provider and return the `err`

appending whatever is the result, allows further debugging over the []reconcilers. 
I decided to imply the less amount of changes in order to avoid unexpected alterations of the controller behavior. Maintainers can provide a direct guideline on this and I'm fine to alter this.

harikrongali replied:
@fblgit you need to check for err in other cases as well. i see error validation partial here. in above conditions like istio/ingress, validation is not done

leoluz replied:
The problem is that the slice is declared as:

```golang
trafficReconcilers := []trafficrouting.TrafficRoutingReconciler{}
```

As `TrafficRoutingReconciler` is an interface, this code could lead in some situation similar to this:

```golang
trafficReconcilers = append(trafficReconcilers, nil)
for _, reconciler := range trafficReconcilers {
    reconciler.SetWeight()
}
```

which will panic with `invalid memory address or nil pointer dereference`

> @fblgit: No matter what u append here if you have err != nil reconcileTrafficRouting will not process any provider and return the err

Regardless if the consumers will verify the error or not, I wouldn't append a `nil` value to a slice to avoid the `nil pointer` error to ever happen.

alexmt replied:
I actually agree with @leoluz . As far as I understand there is no big difference in controller behavior right now but `nil` in the slice might be a surprise for someone in the future. Just to avoid debugging nil dereference error I would not return slice with nil element.

fblgit replied:
change commited.

leoluz commented on rollout/trafficrouting.go:
```diff
@@ -17,152 +17,177 @@ import (
 			Recorder:       c.recorder,
 			ControllerKind: controllerKind,
 		})
+		trafficReconcilers = append(trafficReconcilers, smi_reconcilier)
```
Same here. Maybe its better to check the error first before appending to the `trafficReconcilers`

xavipanda replied:
The errors are controlled here:
```
reconcileTrafficRouting() error {
	reconcilers, err := c.newTrafficRoutingReconciler(c)
	// a return here does ensure that all trafficReconcilers are healthy
	// and sane in syntax
	if err != nil {
		return err
	}
```
the err return is part of the original code, so even in the previous case a `[valid, nil], err` or `[nil], err` or `[valid], err` does result in the same outcome. If `err` is present, the whole reconciliation fails

harikrongali replied:
any traffic controller fails with err, it should return error and fail whole reconciliation .. in the above trafficReconcilers updates, err checks are not done for all controllers.

fblgit replied:
change commited.

fblgit commented:
quite strange...
```
 FAIL: TestFunctionalSuite/TestRolloutAbortRetryPromote (6.41s)
```

But when I run this one locally:
```
=== RUN   TestFunctionalSuite/TestRolloutAbortRetryPromote
time="2021-09-20T16:49:43Z" level=info msg="Deleting e2e-test-name=TestRolloutAbortRetryPromote"
W0920 16:49:43.915482  166675 warnings.go:70] policy/v1beta1 PodDisruptionBudget is deprecated in v1.21+, unavailable in v1.25+; use policy/v1 PodDisruptionBudget
W0920 16:49:43.917669  166675 warnings.go:70] policy/v1beta1 PodDisruptionBudget is deprecated in v1.21+, unavailable in v1.25+; use policy/v1 PodDisruptionBudget
time="2021-09-20T16:49:43Z" level=info msg="Event watcher started"
time="2021-09-20T16:49:44Z" level=info msg="rollout.argoproj.io/abort-retry-promote created\n" rollout=abort-retry-promote
time="2021-09-20T16:49:44Z" level=info msg="Waiting for condition: status=Healthy" rollout=abort-retry-promote
time="2021-09-20T16:49:46Z" level=info msg="Condition 'status=Healthy' met after 2s" rollout=abort-retry-promote
time="2021-09-20T16:49:46Z" level=info msg="Updated rollout pod spec: 2021-09-20T16:49:46.211363209Z" rollout=abort-retry-promote
time="2021-09-20T16:49:46Z" level=info msg="Waiting for condition: status=Paused" rollout=abort-retry-promote
time="2021-09-20T16:49:47Z" level=info msg="Condition 'status=Paused' met after 1s" rollout=abort-retry-promote
time="2021-09-20T16:49:47Z" level=info msg="Pod expectation 'revision:1 pod count == 1' met" rollout=abort-retry-promote
time="2021-09-20T16:49:47Z" level=info msg="Pod expectation 'revision:2 pod count == 1' met" rollout=abort-retry-promote
time="2021-09-20T16:49:47Z" level=info msg="Aborted rollout" rollout=abort-retry-promote
time="2021-09-20T16:49:47Z" level=info msg="Waiting for condition: status=Degraded" rollout=abort-retry-promote
time="2021-09-20T16:49:47Z" level=info msg="Condition 'status=Degraded' met after 0s" rollout=abort-retry-promote
time="2021-09-20T16:49:47Z" level=info msg="Retried rollout" rollout=abort-retry-promote
time="2021-09-20T16:49:47Z" level=info msg="Waiting for condition: status=Paused" rollout=abort-retry-promote
time="2021-09-20T16:49:48Z" level=info msg="Condition 'status=Paused' met after 0s" rollout=abort-retry-promote
time="2021-09-20T16:49:48Z" level=info msg="Pod expectation 'revision:1 pod count == 1' met" rollout=abort-retry-promote
time="2021-09-20T16:49:48Z" level=info msg="Pod expectation 'revision:2 pod count == 1' met" rollout=abort-retry-promote
time="2021-09-20T16:49:48Z" level=info msg="Waiting for condition: status=Healthy" rollout=abort-retry-promote
time="2021-09-20T16:49:51Z" level=info msg="Condition 'status=Healthy' met after 2s" rollout=abort-retry-promote
time="2021-09-20T16:49:51Z" level=info msg="Pod expectation 'revision:1 pod count == 0' met" rollout=abort-retry-promote
time="2021-09-20T16:49:51Z" level=info msg="Pod expectation 'revision:2 pod count == 1' met" rollout=abort-retry-promote
time="2021-09-20T16:49:51Z" level=info msg="Deleting e2e-test-name=TestRolloutAbortRetryPromote"
time="2021-09-20T16:49:51Z" level=info msg="Event watcher stopped" rollout=abort-retry-promote
W0920 16:49:51.153409  166675 warnings.go:70] policy/v1beta1 PodDisruptionBudget is deprecated in v1.21+, unavailable in v1.25+; use policy/v1 PodDisruptionBudget
W0920 16:49:51.155028  166675 warnings.go:70] policy/v1beta1 PodDisruptionBudget is deprecated in v1.21+, unavailable in v1.25+; use policy/v1 PodDisruptionBudget
time="2021-09-20T16:49:51Z" level=info msg="Deleting e2e-test-name"
W0920 16:49:51.194590  166675 warnings.go:70] policy/v1beta1 PodDisruptionBudget is deprecated in v1.21+, unavailable in v1.25+; use policy/v1 PodDisruptionBudget
W0920 16:49:51.196061  166675 warnings.go:70] policy/v1beta1 PodDisruptionBudget is deprecated in v1.21+, unavailable in v1.25+; use policy/v1 PodDisruptionBudget
--- PASS: TestFunctionalSuite (7.69s)
    --- PASS: TestFunctionalSuite/TestRolloutAbortRetryPromote (7.29s)
```

I believe the problem is unrelated to the PR itself, can someone confirm? thank you

huikang reviewed (commented):
Looks great!

fblgit commented on docs/features/traffic-management/mixed.md:
```diff
@@ -0,0 +1,60 @@
+## Requirements
+
+The use of multiple providers requires that both providers comply with its minimum requirements independently.
+By example, if you wan't to use NGiNX and SMI you would need to have both SMI and NGiNX in place and produce the rollout configuration
```
fixed! thank you

fblgit commented:
@khhirani / @harikrongali
Please can you tell me the merge order for this milestone ?
I'm fine to rebase it, resolve conflicts, retest, etc at once in order to avoid further conflicts. I will align this one with your provided timeline and merging order.

jessesuen commented:
Hi @fblgit ! Thank you for this contribution! Regarding “merge order”, theres no real order per-se. It's affected by release timing, or if/when the PR is ready. This past week, I had been prioritizing and only merging things which were required for the v1.1 release candidate, since we are running behind on v1.1. I actually didn't realize @harikrongali had targeted this for v1.1 otherwise I would have paid closer attention. Sorry about that.

Generally speaking, we try to avoid producing unnecessary work for external contributors if it can be avoided, and instead place the burden on project maintainers. This includes holding off on merging big sweeping changes (until impacted smaller PRs get in first), so that outside contributors don't have to deal with the conflicts. Other times we'll contribute to the PR itself to fix conflicts ourselves.

In any case, this PR now on my radar and based on feedback, appears close to complete. So if you rebase now, I think you won't have to deal with conflicts again.

fblgit commented:
Noted and thank you, will be ready before Monday.

Regards

harikrongali commented:
@fblgit It is from my end. Sorry for that, I wanted to get in as we have delayed release. But as there are review comments on changes, it didn't cut the 1.1 release. I should have updated it last week. We wanted to get in the changes as soon as possible after the review process is done to avoid merge conflicts.

harikrongali commented on docs/getting-started/mixed/index.md:
```diff
@@ -0,0 +1,233 @@
+# Getting Started - Multiple Providers (Service Mesh Interface and NGiNX Ingress)
+
+!!! important
+    Available since v1.1
```
can you please update this to 1.2

fblgit replied:
@harikrongali rebased to master, re-tested, etc and looks good against master, u think we ain't in time for 1.1 ? if u are sure we can't make it for 1.1 I will change it for 1.2, please confirm.

harikrongali replied:
yes..1.1 release cut is already made last week and only bugs or regressions will go in..no features...next features will be part of 1.2

fblgit replied:
done

fblgit commented:
Ready.

harikrongali commented on docs/features/traffic-management/mixed.md:
```diff
@@ -0,0 +1,60 @@
+# Multiple Providers
+!!! note
+
+    Multiple trafficRouting is available since Argo Rollouts v1.1
```
please update here as well v1.2

fblgit replied:
done
